### DISCUSSION TITLE
Remote Store

### DISCUSSION THREAD

**Discussion by KaoruDev** [prometheus/client_ruby#263] [Ideas]
I'm evaluating using this gem or [discourse/prometheus_exporter](https://github.com/discourse/prometheus_exporter) I ended up choosing this gem because of its simplicity and the fact that it's backed by the Prometheus community (not a company).

The issues around `DirectFileStore` give me pause. I started to wonder how the discourse gem solved this problem and it looks like it comes baked in with a [mini Pushgate](https://github.com/discourse/prometheus_exporter#multi-process-mode). Instead of keeping stats in the main process, it pushes them to a separate process via `localhost`. 

I like this idea more than having to manage CPU and memory of the main application. I'm working in a K8S cluster so I imagine spinning up a sidecar with a mini push gate process running.

Shooting from the hip, I imagine a new class `RemoteProxyStore` (naming is hard) could be created that implements the store API. `RemoteProxyStore` would send stats over to a configurable address where a web server would listen and save metrics in memory and expose them to the Prometheus server.

Before I create a PR etc, I wanted to get people's thoughts. Thanks for creating this gem!

**Comment by dmagliola** [comment]
Initially, I think this is a very interesting idea and worth exploring. One of the explicit design objectives of the "Stores"  system is to enable use cases like this.

I'm not sure we would incorporate this in the official gem, there are a lot of considerations involved in the approach. We'd have to evaluate it once we see what is involved. But I still think this is worth doing and opening up for others.

One of the advantages of the Stores approach is that it'd be very easy for you to make a gem that depends on this one, and implements this store if it turns out it doesn't quite fit in this one, and we link to it from the README as a suggested viable alternative, etc.

How does that sound?

**Reply by KaoruDev** [reply]
cool sounds reasonable thank you!
